### ISSUE TITLE
[Tensorflow QAT] AttributeError: 'NoneType' object has no attribute 'graph_def'

### ISSUE THREAD

**Issue by peiwenhuang27** [intel/neural-compressor#27]
Environment: Google Colab
LPOT Version: 1.6
Tensorflow Version: Official 2.6.0 (with environment variables set as below)
`TF_ENABLE_ONEDNN_OPTS=1`
`TF_ENABLE_MKL_NATIVE_FORMAT=0`

I basically followed the qat example provided [here](https://github.com/intel/lpot/tree/master/examples/tensorflow/qat_conversion).
I used a pretrained model that is to be annotated with only Conv2D being quantized, and used the annotated model for `model.fit()` for several epochs and saved the model.
After that, I use LPOT ModelConversion to convert the model, and the following error occurs:
```
2021-09-10 03:07:43 [INFO] Pass Quantization elapsed time: 7581.68 ms
2021-09-10 03:07:44 [INFO] Pass FreezeFakeQuantOpOptimizer elapsed time: 283.8 ms
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/lpot/adaptor/tf_utils/graph_converter.py", line 534, in quantize
    self._fuse_requantize_with_fused_quantized_node()
  File "/usr/local/lib/python3.7/dist-packages/lpot/adaptor/tf_utils/graph_converter.py", line 698, in _fuse_requantize_with_fused_quantized_node
    self.device).do_transformation()
  File "/usr/local/lib/python3.7/dist-packages/lpot/adaptor/tf_utils/graph_rewriter/int8/fuse_conv_requantize.py", line 47, in __init__
    self.graph_info = self.graph_analyzer.parse_graph()
  File "/usr/local/lib/python3.7/dist-packages/lpot/adaptor/tf_utils/graph_rewriter/graph_util.py", line 611, in parse_graph
    each_input)].outputs.append(node_name)
KeyError: 'model_3/quant_31/StatefulPartitionedCall/StatefulPartitionedCall/MovingAvgQuantize/FakeQuantWithMinMaxVars'
2021-09-10 03:07:44 [ERROR] Fail to quantize graph due to 'model_3/quant_31/StatefulPartitionedCall/StatefulPartitionedCall/MovingAvgQuantize/FakeQuantWithMinMaxVars'.
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-515087c4513a> in <module>()
      4 conversion.destination = 'default'
      5 conversion.model = common.Model('./q_aware_model')
----> 6 q_model = conversion()
      7 q_model.save('quantized_model')

2 frames
/usr/local/lib/python3.7/dist-packages/lpot/experimental/model_conversion.py in __call__(self)
     94 
     95         self.adaptor = FRAMEWORKS[self.framework](framework_specific_info)
---> 96         q_model = self.adaptor.convert(self._model, self._source, self._destination)
     97 
     98         # when eval_func is None but metric or _eval_dataloader is set by yaml or code,

/usr/local/lib/python3.7/dist-packages/lpot/adaptor/tensorflow.py in convert(self, model, source, destination)
    814                                    fake_quant=True)
    815 
--> 816         return converter.convert()
    817 
    818     @dump_elapsed_time("Pass recover model")

/usr/local/lib/python3.7/dist-packages/lpot/adaptor/tf_utils/graph_converter.py in convert(self)
    247         if len(self.bf16_ops) > 0:
    248             model = self.bf16_convert()
--> 249         post_cse_graph_def = PostCseOptimizer(model.graph_def).do_transformation()
    250         post_cse_graph_def.library.CopyFrom(self.model.graph_def.library)
    251         model.graph_def = post_cse_graph_def

AttributeError: 'NoneType' object has no attribute 'graph_def'
```

My original code (simplified):
```
model = tf.keras.models.load_model('model')

import tensorflow_model_optimization as tfmot

def apply_quantization_to_Conv2D(layer):
  if isinstance(layer, tf.keras.layers.Conv2D):
    return tfmot.quantization.keras.quantize_annotate_layer(layer)
  return layer

annotated_model = tf.keras.models.clone_model(model, clone_function=apply_quantization_to_Conv2D)

q_aware_model = tfmot.quantization.keras.quantize_apply(annotated_model)
q_aware_model.summary()

q_aware_model.compile(optimizer='adam', loss='mse')
q_aware_model.fit(x=[X_q, X_norm_q], y=y_q, 
                  batch_size=64,
                  epochs=45)
q_aware_model.save('./q_aware_model')

from lpot.experimental import ModelConversion, common
conversion = ModelConversion()
conversion.source = 'QAT'
conversion.destination = 'default'
conversion.model = common.Model('./q_aware_model')
q_model = conversion()
q_model.save('quantized_model')
```

Please find model [here](https://drive.google.com/drive/folders/1NXJiCpIW26uL6_5VfwGBXf-hnWFA1_LX?usp=sharing).
Thanks!

**Comment by peiwenhuang27** [comment]
Hi, if possible, could someone take a look at this issue? I'm trying to meet a project deadline, and this is kind of a crucial bottleneck I'm facing. Your help is much appreciated, thanks!

**Comment by ftian1** [comment]
sure. we will check it asap

**Comment by peiwenhuang27** [comment]
Thank you so much!
By the way, the model link above is the pretrained version, for quantization aware trained version please check [this one](https://drive.google.com/drive/folders/1-yBAwQQundm87p73w9845jduxIOEpXVs?usp=sharing)

**Comment by guomingz** [comment]
@peiwenhuang27  could you please provide the full script to reproduce this issue？
The above script is not completed.

**Comment by peiwenhuang27** [comment]
Sorry, I may not be able to give the training data used in the original script, however, since the quantization aware training part is successfully completed, perhaps you could directly use the quantization aware trained version model provided [here](https://drive.google.com/drive/folders/1-yBAwQQundm87p73w9845jduxIOEpXVs?usp=sharing)
Then the script is simply just
```
from lpot.experimental import ModelConversion, common
conversion = ModelConversion()
conversion.source = 'QAT'
conversion.destination = 'default'
conversion.model = common.Model('./q_aware_model')
q_model = conversion()
q_model.save('quantized_model')
```
I checked the pretrained model, the node name triggering `KeyError` actually does exist in the original `conversion.model.graph`, I am not sure if something went wrong in `graph_util.py", line 611, in parse_graph`

**Comment by peiwenhuang27** [comment]
I found out that the last time `parse_graph` was called (as in the error log above), the node name triggering `KeyError` does not exist in the passed `graph_def` object, and looks like the node name differs in each execution, perhaps due to random seed not set.

**Comment by Zhiwei35** [comment]
@peiwenhuang27 hi , I cant open your quantization aware trained version model [link](https://drive.google.com/drive/folders/1-yBAwQQundm87p73w9845jduxIOEpXVs?usp=sharing) due to some internet issues. Can you send me the model by email or other ways I can get?

**Comment by Zhiwei35** [comment]
> @guomingz After look at the source code, I found that
> In `graph_converter.py def _fuse_requantize_with_fused_quantized_node(self): line 691`:
> I found out that before
> 
> ```
> if self.fake_quant:
>             self._tmp_graph_def = FreezeFakeQuantOpOptimizer(
>                 self._tmp_graph_def).do_transformation()
> ```
> 
> the node still exists in the graph, but after transformation, the node disappears from the graph, but is still referenced as input by some other nodes in the graph. Thus, in `FuseConvRequantizeTransformer def __init__()`, when initializing with `parse_graph`, the following
> 
> ```
> for node_name, node_details in self.node_name_details.items():
>             # update the upper node's output infomation.
>             for each_input in node_details.node.input:
>                 self.node_name_details[GraphRewriterHelper.node_name_from_input(
>                     each_input)].outputs.append(node_name)
> ```
> 
> will trigger a `KeyError`
@peiwenhuang27 OK，I get your point and I will reproduce the issue first, then give you a response

**Comment by Zhiwei35** [comment]
@peiwenhuang27 hi, man, i use your code and model named q_aware_model, i didn't get the same log as you. The code i use is below:
`from lpot.experimental import ModelConversion, common
conversion = ModelConversion()
conversion.source = 'QAT'
conversion.destination = 'default'
conversion.model = common.Model('q_aware_model/q_aware_model')
q_model = conversion()
q_model.save('quantized_model')`
the link i got the model is [here](https://drive.google.com/drive/folders/1-yBAwQQundm87p73w9845jduxIOEpXVs)
the log of mine is:
`
2021-09-14 14:32:58 [INFO] Restoring parameters from q_aware_model/q_aware_model/variables/variables
2021-09-14 14:32:59 [INFO] Pass PostCseOptimizer elapsed time: 34.23 ms
2021-09-14 14:33:00.923046: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
2021-09-14 14:33:01 [WARNING] From /home2/zhiweihu/anaconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/saved_model/signature_def_utils_impl.py:201: build_tensor_info (from tensorflow.python.saved_model.utils_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.utils.build_tensor_info or tf.compat.v1.saved_model.build_tensor_info.
2021-09-14 14:33:01 [INFO] No assets to save.
2021-09-14 14:33:01 [INFO] No assets to write.
2021-09-14 14:33:02 [INFO] SavedModel written to: /home2/zhiweihu/work/quantized_model/saved_model.pb
2021-09-14 14:33:02 [INFO] Save quantized model to /home2/zhiweihu/work/quantized_model.
`
please check your script and send us the right one that can reproduce the issue, thx

**Comment by Zhiwei35** [comment]
@peiwenhuang27 although i cant reproduced your issue, i think you can try to modify the line 154 of tf_utils/graph_rewriter/int8/freeze_fake_quant.py from

> return GraphAnalyzer().dump_graph()

to

> self.graph_analyzer.dump_graph()

**Comment by peiwenhuang27** [comment]
Here is a standalone script for reproducing the issue, please check: https://colab.research.google.com/drive/106Lt4mNzHWoCD56KtSPPInD6KNbxgkwV?usp=sharing
I have tried your suggestion
> return self.graph_analyzer.dump_graph()

unfortunately it does not work.

**Comment by guomingz** [comment]
i got the similar result as zhiwei mentioned in 

> @peiwenhuang27 hi, man, i use your code and model named q_aware_model, i didn't get the same log as you. The code i use is below:
> 
> > from lpot.experimental import ModelConversion, common
> > conversion = ModelConversion()
> > conversion.source = 'QAT'
> > conversion.destination = 'default'
> > conversion.model = common.Model('q_aware_model/q_aware_model')
> > q_model = conversion()
> > q_model.save('quantized_model')
> 
> the link i got the model is [here](https://drive.google.com/drive/folders/1-yBAwQQundm87p73w9845jduxIOEpXVs)
> the log of mine is:
> 
> > 2021-09-14 14:32:58 [INFO] Restoring parameters from q_aware_model/q_aware_model/variables/variables
> > 2021-09-14 14:32:59 [INFO] Pass PostCseOptimizer elapsed time: 34.23 ms
> > 2021-09-14 14:33:00.923046: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
> > 2021-09-14 14:33:01 [WARNING] From /home2/zhiweihu/anaconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/saved_model/signature_def_utils_impl.py:201: build_tensor_info (from tensorflow.python.saved_model.utils_impl) is deprecated and will be removed in a future version.
> > Instructions for updating:
> > This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.utils.build_tensor_info or tf.compat.v1.saved_model.build_tensor_info.
> > 2021-09-14 14:33:01 [INFO] No assets to save.
> > 2021-09-14 14:33:01 [INFO] No assets to write.
> > 2021-09-14 14:33:02 [INFO] SavedModel written to: /home2/zhiweihu/work/quantized_model/saved_model.pb
> > 2021-09-14 14:33:02 [INFO] Save quantized model to /home2/zhiweihu/work/quantized_model.
> 
> please check your script and send us the right one that can reproduce the issue, thx

i got the same result as zhiwei reported here.
@peiwenhuang27  how do we reproduce your issue w/o full script?

**Comment by guomingz** [comment]
> Here is a standalone script for reproducing the issue, please check: https://colab.research.google.com/drive/106Lt4mNzHWoCD56KtSPPInD6KNbxgkwV?usp=sharing
> I have tried your suggestion
> 
> > return self.graph_analyzer.dump_graph()
> 
> unfortunately it does not work.

from lpot.experimental import ModelConversion, common
conversion = ModelConversion()
conversion.source = 'QAT'
conversion.destination = 'default'
**conversion.model = common.Model('/content/q_aware_model/')**
q_model = conversion()
q_model.save('quantized_model')

==============
/content/q_aware_model/ is equal to https://drive.google.com/drive/folders/1-yBAwQQundm87p73w9845jduxIOEpXVs ?

**Comment by peiwenhuang27** [comment]
@guomingz yes, it is the same model, I only compressed it to zip format to be easily accessed in the script.

**Comment by Zhiwei35** [comment]
> Here is a standalone script for reproducing the issue, please check: https://colab.research.google.com/drive/106Lt4mNzHWoCD56KtSPPInD6KNbxgkwV?usp=sharing
> I have tried your suggestion
> 
> > return self.graph_analyzer.dump_graph()
> 
> unfortunately it does not work.

hi ， the log I pasted yesterday is what I print in intel-tensorflow 2.5.0 with TF_ENABLE_MKL_NATIVE_FORMAT=0, you can try with that first.

**Comment by peiwenhuang27** [comment]
I see, it does work with intel-tensorflow 2.5.0. I thought LPOT can be directly run on official Tensorflow 2.6.0 without intel-tensorflow? It appears there still is some discrepancy between the two versions.
Thanks anyway!

**Comment by peiwenhuang27** [comment]
Though the conversion was successful, I think there may still be some issue.
Using intel-tensorflow 2.5, my log is:
```
2021-09-16 01:02:07 [INFO] Generating grammar tables from /usr/lib/python3.7/lib2to3/Grammar.txt
2021-09-16 01:02:07 [INFO] Generating grammar tables from /usr/lib/python3.7/lib2to3/PatternGrammar.txt
WARNING:tensorflow:From /content/lpot/lpot/model/model.py:501: load (from tensorflow.python.saved_model.loader_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.loader.load or tf.compat.v1.saved_model.load. There will be a new function for importing SavedModels in Tensorflow 2.0.
2021-09-16 01:02:13 [WARNING] From /content/lpot/lpot/model/model.py:501: load (from tensorflow.python.saved_model.loader_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.loader.load or tf.compat.v1.saved_model.load. There will be a new function for importing SavedModels in Tensorflow 2.0.
INFO:tensorflow:Restoring parameters from /content/q_aware_model/variables/variables
2021-09-16 01:02:13 [INFO] Restoring parameters from /content/q_aware_model/variables/variables
2021-09-16 01:02:20 [INFO] ConvertLayoutOptimizer elapsed time: 0.18 ms
2021-09-16 01:02:21 [INFO] Pass GrapplerOptimizer elapsed time: 1279.76 ms
2021-09-16 01:02:21 [INFO] Pass RemoveTrainingNodesOptimizer elapsed time: 30.11 ms
2021-09-16 01:02:21 [INFO] Pass SplitSharedInputOptimizer elapsed time: 5.48 ms
2021-09-16 01:02:21 [INFO] Pass GraphFoldConstantOptimizer elapsed time: 1.39 ms
2021-09-16 01:02:21 [INFO] Pass FuseColumnWiseMulOptimizer elapsed time: 1.01 ms
WARNING:tensorflow:From /content/lpot/lpot/adaptor/tf_utils/util.py:318: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.compat.v1.graph_util.extract_sub_graph`
2021-09-16 01:02:21 [WARNING] From /content/lpot/lpot/adaptor/tf_utils/util.py:318: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.compat.v1.graph_util.extract_sub_graph`
2021-09-16 01:02:21 [INFO] Pass StripUnusedNodesOptimizer elapsed time: 10.41 ms
2021-09-16 01:02:21 [INFO] Pass GraphCseOptimizer elapsed time: 3.19 ms
2021-09-16 01:02:21 [INFO] Pass FoldBatchNormNodesOptimizer elapsed time: 1.65 ms
2021-09-16 01:02:21 [INFO] Pass UpdateEnterOptimizer elapsed time: 1.06 ms
2021-09-16 01:02:21 [INFO] Pass ConvertLeakyReluOptimizer elapsed time: 1.52 ms
2021-09-16 01:02:21 [INFO] Pass InjectDummyBiasAddOptimizer elapsed time: 1.79 ms
2021-09-16 01:02:21 [INFO] Pass ConvertAddToBiasAddOptimizer elapsed time: 2.39 ms
2021-09-16 01:02:21 [INFO] Pass FuseTransposeReshapeOptimizer elapsed time: 2.46 ms
2021-09-16 01:02:21 [INFO] Pass FuseConvWithMathOptimizer elapsed time: 2.74 ms
2021-09-16 01:02:22 [INFO] Pass Pre Optimization elapsed time: 8008.38 ms
2021-09-16 01:02:35 [WARNING] Found possible input node names: ['quantize_layer/quantize_layer_min', 'quantize_layer/quantize_layer_max', '1/kernel', 'quant_1/kernel_min', 'quant_1/kernel_max', '1/bias', 'quant_1/post_activation_min', 'quant_1/post_activation_max', '2/kernel', 'quant_2/kernel_min', 'quant_2/kernel_max', '2/bias', 'quant_2/post_activation_min', 'quant_2/post_activation_max', '32/kernel', 'quant_32/kernel_min', 'quant_32/kernel_max', '32/bias', 'quant_32/post_activation_min', 'quant_32/post_activation_max', '42/kernel', 'quant_42/kernel_min', 'quant_42/kernel_max', '42/bias', 'quant_42/post_activation_min', 'quant_42/post_activation_max', '52/kernel', 'quant_52/kernel_min', 'quant_52/kernel_max', '52/bias', 'quant_52/post_activation_min', 'quant_52/post_activation_max', '31/kernel', 'quant_31/kernel_min', 'quant_31/kernel_max', '31/bias', 'quant_31/post_activation_min', 'quant_31/post_activation_max', 'LSTM_2/lstm_cell_4/kernel', 'LSTM_2/lstm_cell_4/recurrent_kernel', 'LSTM_2/lstm_cell_4/bias', '41/kernel', 'quant_41/kernel_min', 'quant_41/kernel_max', '41/bias', 'quant_41/post_activation_min', 'quant_41/post_activation_max', 'dense_20/kernel', 'dense_20/bias', '51/kernel', 'quant_51/kernel_min', 'quant_51/kernel_max', '51/bias', 'quant_51/post_activation_min', 'quant_51/post_activation_max', 'dense_21/kernel', 'dense_21/bias', 'dense_22/kernel', 'dense_22/bias', 'LSTM_1/lstm_cell_5/kernel', 'LSTM_1/lstm_cell_5/recurrent_kernel', 'LSTM_1/lstm_cell_5/bias', 'dense_23/kernel', 'dense_23/bias', 'dense_18/kernel', 'dense_18/bias', 'dense_19/kernel', 'dense_19/bias', 'serving_default_input_noisy', 'serving_default_input_noisy_norm'], output node names: ['StatefulPartitionedCall'].
2021-09-16 01:02:46 [WARNING] Found possible input node names: ['quantize_layer/quantize_layer_min', 'quantize_layer/quantize_layer_max', '1/kernel', 'quant_1/kernel_min', 'quant_1/kernel_max', '1/bias', 'quant_1/post_activation_min', 'quant_1/post_activation_max', '2/kernel', 'quant_2/kernel_min', 'quant_2/kernel_max', '2/bias', 'quant_2/post_activation_min', 'quant_2/post_activation_max', '32/kernel', 'quant_32/kernel_min', 'quant_32/kernel_max', '32/bias', 'quant_32/post_activation_min', 'quant_32/post_activation_max', '42/kernel', 'quant_42/kernel_min', 'quant_42/kernel_max', '42/bias', 'quant_42/post_activation_min', 'quant_42/post_activation_max', '52/kernel', 'quant_52/kernel_min', 'quant_52/kernel_max', '52/bias', 'quant_52/post_activation_min', 'quant_52/post_activation_max', '31/kernel', 'quant_31/kernel_min', 'quant_31/kernel_max', '31/bias', 'quant_31/post_activation_min', 'quant_31/post_activation_max', 'LSTM_2/lstm_cell_4/kernel', 'LSTM_2/lstm_cell_4/recurrent_kernel', 'LSTM_2/lstm_cell_4/bias', '41/kernel', 'quant_41/kernel_min', 'quant_41/kernel_max', '41/bias', 'quant_41/post_activation_min', 'quant_41/post_activation_max', 'dense_20/kernel', 'dense_20/bias', '51/kernel', 'quant_51/kernel_min', 'quant_51/kernel_max', '51/bias', 'quant_51/post_activation_min', 'quant_51/post_activation_max', 'dense_21/kernel', 'dense_21/bias', 'dense_22/kernel', 'dense_22/bias', 'LSTM_1/lstm_cell_5/kernel', 'LSTM_1/lstm_cell_5/recurrent_kernel', 'LSTM_1/lstm_cell_5/bias', 'dense_23/kernel', 'dense_23/bias', 'dense_18/kernel', 'dense_18/bias', 'dense_19/kernel', 'dense_19/bias', 'serving_default_input_noisy', 'serving_default_input_noisy_norm'], output node names: ['StatefulPartitionedCall'].
INFO:tensorflow:Restoring parameters from /content/q_aware_model/variables/variables
2021-09-16 01:02:48 [INFO] Restoring parameters from /content/q_aware_model/variables/variables
2021-09-16 01:02:49 [INFO] Pass PostCseOptimizer elapsed time: 28.34 ms
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/saved_model/signature_def_utils_impl.py:201: build_tensor_info (from tensorflow.python.saved_model.utils_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.utils.build_tensor_info or tf.compat.v1.saved_model.build_tensor_info.
2021-09-16 01:02:51 [WARNING] From /usr/local/lib/python3.7/dist-packages/tensorflow/python/saved_model/signature_def_utils_impl.py:201: build_tensor_info (from tensorflow.python.saved_model.utils_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.utils.build_tensor_info or tf.compat.v1.saved_model.build_tensor_info.
INFO:tensorflow:No assets to save.
2021-09-16 01:02:51 [INFO] No assets to save.
INFO:tensorflow:No assets to write.
2021-09-16 01:02:51 [INFO] No assets to write.
INFO:tensorflow:SavedModel written to: /content/lpot/quantized_model/saved_model.pb
2021-09-16 01:02:51 [INFO] SavedModel written to: /content/lpot/quantized_model/saved_model.pb
2021-09-16 01:02:51 [INFO] Save quantized model to /content/lpot/quantized_model.
```

However, I think the quantized variables were not saved after conversion. Originally my saved model has a file `variables.data-00000-of-00001` which has size of 102 MB, but the quantized saved model only has a `saved_model.pb` with 2.39 MB. Also, when I load the model for inference (also under intel-tensorflow 2.5), it shows the following error message:
```
FailedPreconditionError: Could not find variable quant_2/post_activation_max. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status=Not found: Container localhost does not exist. (Could not find resource: localhost/quant_2/post_activation_max)
	 [[{{node model_3/quant_2/MovingAvgQuantize/FakeQuantWithMinMaxVars/ReadVariableOp_1}}]]

```

**Comment by guomingz** [comment]
The above msg is not make sense, especially on  `2021-09-16 01:02:35 [WARNING] Found possible input node names: ['quantize_layer/quantize_layer_min', 'quantize_layer/quantize_layer_max', '1/kernel', 'quant_1/kernel_min', 'quant_1/kernel_max', '1/bias', 'quant_1/post_activation_min', 'quant_1/post_activation_max', '2/kernel', 'quant_2/kernel_min', 'quant_2/kernel_max', '2/bias', 'quant_2/post_activation_min', 'quant_2/post_activation_max', '32/kernel', 'quant_32/kernel_min', 'quant_32/kernel_max', '32/bias', 'quant_32/post_activation_min', 'quant_32/post_activation_max', '42/kernel', 'quant_42/kernel_min', 'quant_42/kernel_max', '42/bias', 'quant_42/post_activation_min', 'quant_42/post_activation_max', '52/kernel', 'quant_52/kernel_min', 'quant_52/kernel_max', '52/bias', 'quant_52/post_activation_min', 'quant_52/post_activation_max', '31/kernel', 'quant_31/kernel_min', 'quant_31/kernel_max', '31/bias', 'quant_31/post_activation_min', 'quant_31/post_activation_max', 'LSTM_2/lstm_cell_4/kernel', 'LSTM_2/lstm_cell_4/recurrent_kernel', 'LSTM_2/lstm_cell_4/bias', '41/kernel', 'quant_41/kernel_min', 'quant_41/kernel_max', '41/bias', 'quant_41/post_activation_min', 'quant_41/post_activation_max', 'dense_20/kernel', 'dense_20/bias', '51/kernel', 'quant_51/kernel_min', 'quant_51/kernel_max', '51/bias', 'quant_51/post_activation_min', 'quant_51/post_activation_max', 'dense_21/kernel', 'dense_21/bias', 'dense_22/kernel', 'dense_22/bias', 'LSTM_1/lstm_cell_5/kernel', 'LSTM_1/lstm_cell_5/recurrent_kernel', 'LSTM_1/lstm_cell_5/bias', 'dense_23/kernel', 'dense_23/bias', 'dense_18/kernel', 'dense_18/bias', 'dense_19/kernel', 'dense_19/bias', 'serving_default_input_noisy', 'serving_default_input_noisy_norm'], output node names: ['StatefulPartitionedCall'].`

what i got is `2021-09-16 13:09:08 [WARNING] Found possible input node names: ['input_noisy', 'input_noisy_norm'], output node names: ['outputMask'].`

I'm using the intel-tensorflow 2.6.0 and the attachment is the model converted from my side.
[saved_model.zip](https://github.com/intel/neural-compressor/files/7175073/saved_model.zip)

**Comment by peiwenhuang27** [comment]
Using inetl-tensorflow 2.6.0 with `TF_ENABLE_ONEDNN_OPTS=1, TF_ENABLE_MKL_NATIVE_FORMAT=0`, I get the same KeyError as originally reported here.
Also, my model does not have the log `2021-09-14 14:32:58 [INFO] Restoring parameters from q_aware_model/q_aware_model/variables/variables` from @Zhiwei35, could it be possible that you are using saved_model_session while mine is detected as a keras_session?
May I ask what version of LPOT you use? I am using the binary version via `pip install lpot` (also tried building from git clone and got the same result).

**Comment by guomingz** [comment]
i'm using lpot internal master branch.
And i printed the model type which is  TensorflowSavedModelModel.

**Comment by Zhiwei35** [comment]
@peiwenhuang27 I used **LPOT v1.6 with pip install lpot** and intel-tensorflow 2.6.0 to run your script just now. And I got the same result and model as Guoming pasted above. Our logs are:

> 2021-09-16 15:20:01 [INFO] Pass StripUnusedNodesOptimizer elapsed time: 215.19 ms
> 2021-09-16 15:20:02 [INFO] Pass GraphCseOptimizer elapsed time: 55.22 ms
> 2021-09-16 15:20:02 [INFO] Pass FoldBatchNormNodesOptimizer elapsed time: 53.03 ms
> 2021-09-16 15:20:02 [INFO] Pass UpdateEnterOptimizer elapsed time: 51.38 ms
> 2021-09-16 15:20:02 [INFO] Pass ConvertLeakyReluOptimizer elapsed time: 52.53 ms
> 2021-09-16 15:20:02 [INFO] Pass InjectDummyBiasAddOptimizer elapsed time: 55.38 ms
> 2021-09-16 15:20:02 [INFO] Pass ConvertAddToBiasAddOptimizer elapsed time: 52.79 ms
> 2021-09-16 15:20:02 [INFO] Pass FuseTransposeReshapeOptimizer elapsed time: 52.62 ms
> 2021-09-16 15:20:02 [INFO] Pass FuseConvWithMathOptimizer elapsed time: 52.61 ms
> 2021-09-16 15:20:02 [WARNING] Node name unused_control_flow_input_47 specified in yaml doesn't exist in the model.
> 2021-09-16 15:20:02 [WARNING] Found possible input node names: ['input_noisy', 'input_noisy_norm'], output node names: ['outputMask'].
> 2021-09-16 15:20:02 [INFO] Pass Pre Optimization elapsed time: 5291.59 ms
> 2021-09-16 15:20:02 [WARNING] Node name unused_control_flow_input_47 specified in yaml doesn't exist in the model.
> 2021-09-16 15:20:02 [WARNING] Found possible input node names: ['input_noisy', 'input_noisy_norm'], output node names: ['outputMask'].
> 2021-09-16 15:20:10 [WARNING] Found possible input node names: ['input_noisy', 'input_noisy_norm'], output node names: ['outputMask'].
> 2021-09-16 15:20:17 [WARNING] Found possible input node names: ['input_noisy', 'input_noisy_norm'], output node names: ['outputMask'].
> 2021-09-16 15:20:21.749364: I tensorflow/core/grappler/devices.cc:75] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0 (Note: TensorFlow was not compiled with CUDA or ROCm support)
> 2021-09-16 15:20:21.749487: I tensorflow/core/grappler/clusters/single_machine.cc:357] Starting new session
> 2021-09-16 15:20:21.777931: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:1137] Optimization results for grappler item: graph_to_optimize
>   function_optimizer: Graph size after: 631 nodes (370), 802 edges (528), time = 7.764ms.
>   function_optimizer: function_optimizer did nothing. time = 0.219ms.
> 
> 2021-09-16 15:20:22.407978: I tensorflow/core/grappler/devices.cc:75] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0 (Note: TensorFlow was not compiled with CUDA or ROCm support)
> 2021-09-16 15:20:22.408114: I tensorflow/core/grappler/clusters/single_machine.cc:357] Starting new session
> 2021-09-16 15:20:22.575757: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:1137] Optimization results for grappler item: tf_graph
>   constant_folding: Graph size after: 422 nodes (-88), 506 edges (-96), time = 61.435ms.
>   constant_folding: Graph size after: 422 nodes (0), 506 edges (0), time = 14.949ms.
> 
> 2021-09-16 15:20:22.968985: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
> 2021-09-16 15:20:22 [WARNING] From /home2/zhiweihu/anaconda3/envs/inteltf26/lib/python3.6/site-packages/tensorflow/python/util/dispatch.py:206: quantize_v2 (from tensorflow.python.ops.array_ops) is deprecated and will be removed after 2017-10-25.
> Instructions for updating:
> `tf.quantize_v2` is deprecated, please use `tf.quantization.quantize` instead.
> 2021-09-16 15:20:23 [INFO] Not find expected types on inputs Const, Const.
> 2021-09-16 15:20:24 [INFO] Pass Quantization elapsed time: 1126.1 ms
> 2021-09-16 15:20:24 [INFO] Pass FreezeFakeQuantOpOptimizer elapsed time: 32.31 ms
> 2021-09-16 15:20:24 [INFO] Pass StripUnusedNodesOptimizer elapsed time: 96.96 ms
> 2021-09-16 15:20:24 [INFO] Pass RemoveTrainingNodesOptimizer elapsed time: 32.91 ms
> 2021-09-16 15:20:24 [INFO] Pass FoldBatchNormNodesOptimizer elapsed time: 32.82 ms
> 2021-09-16 15:20:24 [INFO] Pass MetaOpOptimizer elapsed time: 30.93 ms
> 2021-09-16 15:20:24 [WARNING] Node name unused_control_flow_input_29 specified in yaml doesn't exist in the model.
> 2021-09-16 15:20:24 [WARNING] Found possible input node names: ['input_noisy', 'input_noisy_norm'], output node names: ['outputMask'].
> 2021-09-16 15:20:25 [INFO] Pass PostCseOptimizer elapsed time: 1267.08 ms
> 2021-09-16 15:20:25 [WARNING] From /home2/zhiweihu/anaconda3/envs/inteltf26/lib/python3.6/site-packages/tensorflow/python/saved_model/signature_def_utils_impl.py:201: build_tensor_info (from tensorflow.python.saved_model.utils_impl) is deprecated and will be removed in a future version.
> Instructions for updating:
> This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.utils.build_tensor_info or tf.compat.v1.saved_model.build_tensor_info.
> 2021-09-16 15:20:25 [INFO] No assets to save.
> 2021-09-16 15:20:25 [INFO] No assets to write.
> 2021-09-16 15:20:25 [INFO] SavedModel written to: /home2/zhiweihu/work/quantized_model_lpot16/saved_model.pb
> 2021-09-16 15:20:25 [INFO] Save quantized model to /home2/zhiweihu/work/quantized_model_lpot16.

model we used is [here ](https://drive.google.com/drive/folders/1-yBAwQQundm87p73w9845jduxIOEpXVs). I guess maybe your used a WRONG model.

**Comment by peiwenhuang27** [comment]
Did you try running the colab file DIRECTLY, and not in your local environment? I am sure the model is the same one, the filename is unique in my file system, so there is no way I could have mistaken the model.

**Comment by guomingz** [comment]
We didn't run it on colab. You may execute the code w/o qat process but conversion code only.

**Comment by peiwenhuang27** [comment]
Yes, I understand that the conversion code can be run independently. The model I shared is quantization-aware-trained, so the colab script I gave only contains conversion code. I just suspect the environment could be the key issue here.
By using intel-tensorflow 2.5 & 2.6, should I set `TF_ENABLE_ONEDNN_OPTS=1`
`TF_ENABLE_MKL_NATIVE_FORMAT=0` both or just `TF_ENABLE_MKL_NATIVE_FORMAT` will suffice? Currently I set both.
